core/src/main/java/com/linecorp/armeria/internal/common/AbstractKeepAliveHandler.java

```diff
@@ -109,6 +113,15 @@ protected AbstractKeepAliveHandler(Channel channel, String name, Timer keepAlive
         } else {
             maxConnectionAgeNanos = TimeUnit.MILLISECONDS.toNanos(maxConnectionAgeMillis);
         }
+
+        numActiveConnections = new AtomicInteger();
+        if (isServer) {
+            meterRegistry.gauge("armeria.server.connections", this.numActiveConnections,
+                                AtomicInteger::get);
+        } else {
+            meterRegistry.gauge("armeria.client.connections", this.numActiveConnections,
+                                AtomicInteger::get);
```

Comment:
I don't see that this shows the number of active connections correctly because a `KeepAliveHandler` is created per connection/Channel.
How about creating a shared gauge externally and using it?

Reply:
> I don't see that this correctly shows the number of active connections because a `KeepAliveHandler` is created per connection/Channel.
> How about creating a shared gauge externally and using it?

Hi @ikhoon , thanks for the comments! I missed that `KeepAliveHandler` is created per channel, what I'm thinking about is to make `numActiveConnections` `static`, but I'm not sure how "creating a shared gauge externally and using it" can help to "unify how we count the active connections for client and server" because then we need to create the meter again and again in where `KeepAliveHandler` is constructed.

Reply:
We manage the connection pool in `HttpChannelPool`.
We call `listener.connectionOpen` and `listener.connectionClosed` when a connection is open and closed.
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java#L490
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java#L533
How about also managing the number of connections there? We could probably:
- Add an `AtomicInteger` to `HttpClientFactory`
- Create the gauge in `HttpClientFactory`
- After calling `listener.connectionOpen()`, increase the `AtomicInteger`.
- After calling `listener.connectionClosed()`, decrease the `AtomicInteger`.

For increasing and decreasing the numbers,
- We can define a method `HttpClientFactory.increaseConnection()`
- Or We might:
  - Implement `ConnectionPoolListener.andThen(other)`
  - Make `HttpClientFactory` implement `ConnectionPoolListener`
  - And just call `listener.connectionOpen` and `listener.connectionClosed`

Reply:
> We manage the connection pool in HttpChannelPool.

I think this is probably a better idea.
I think another point is that if there are multiple `ClientFactory`s in play, then it is likely that this metric is reported for only a single arbitrary `ClientFactory`. Implementing this at the `ConnectionPoolListener` level could make it easier to accept a tag (like `name`) which distinguishes between each `ClientFactory`.

Also linking an issue on server-side `ConnectionPoolListener` #3722

Reply:
That's a good point. I thought we might take `MeterIdPrefix` to `ClientFactoryBuilder` so that the meter name could be differentiated.

Reply:
Thank you all for the awesome hint! Really appreciate, I will try the ways your suggested. 🙇

Reply:
Thanks, @kezhenxu94!
Had a chat with @trustin and we have agreed on that we don't have to unify how we count the active connections for client and server. 😄